Classes/EventListener/Linkvalidator/BeforeRecordIsAnalyzedEventListener.php

```diff
@@ -0,0 +1,126 @@
+{
+    protected BrokenLinkRepository $brokenLinkRepository;
+    protected SoftReferenceParserFactory $softReferenceParserFactory;
+    public function __construct(
+        BrokenLinkRepository $brokenLinkRepository,
+        SoftReferenceParserFactory $softReferenceParserFactory
+    ) {
+        $this->brokenLinkRepository = $brokenLinkRepository;
+        $this->softReferenceParserFactory = $softReferenceParserFactory;
+    }
+
+    private function addItemToBrokenLinkRepository(string $table, string $field, array $newsItem, string $foundUrl, string $forbiddenDomain): void
+    {
+        $link = [
+            'record_uid' => $newsItem['uid'],
+            'record_pid' => $newsItem['pid'],
+            'language' => $newsItem['sys_language_uid'],
+            'field' => $field,
+            'table_name' => $table,
+            'url' => $foundUrl,
+            'last_check' => time(),
+            'link_type' => 'external',
+        ];
+        $this->brokenLinkRepository->addBrokenLink($link, false, [
+            'errorType' => 'exception',
+            'exception' => 'Do not link externally to ' . $forbiddenDomain,
+            'errno' => 42, ]);
+    }
+
+    private function findAllParsers(array $conf): iterable
+    {
+        return $this->softReferenceParserFactory->getParsersBySoftRefParserList($conf['softref'], ['subst']);
+    }
+
+    public function __invoke(BeforeRecordIsAnalyzedEvent $event): void
+    {
+        $table = $event->getTableName();
+        $forbiddenDomain = 'example.org';
+        if ($table !== 'tx_news_domain_model_news') {
+            return;
+        }
+        $results = $event->getResults();
+        $newsItem = $event->getRecord();
+        if (!str_contains((string)$newsItem['bodytext'], $forbiddenDomain)) {
+            return;
+        }
+        $field = 'bodytext';
+        $conf = $GLOBALS['TCA']['tx_news_domain_model_news']['columns']['bodytext']['config'];
+        $results = $this->parseField(
+            $conf,
+            $table,
+            $field,
+            $newsItem,
+            $forbiddenDomain,
+            $results
+        );
+        $event->setResults($results);
+    }
+
+    private function parseField(
+        array $conf,
+        string $table,
+        string $field,
+        array $newsItem,
+        string $forbiddenDomain,
+        array $results
+    ): array {
+        foreach ($this->findAllParsers($conf) as $softReferenceParser) {
+            $parserResult = $softReferenceParser->parse(
+                $table,
+                $field,
+                $newsItem['uid'],
+                $newsItem['bodytext']
+            );
+            if (!$parserResult->hasMatched()) {
+                continue;
+            }
+            foreach ($parserResult->getMatchedElements() as $matchedElement) {
+                if (!isset($matchedElement['subst'])) {
+                    continue;
+                }
+                $results = $this->matchUrl(
```

Comment:
The previous results in the loop are overwritten, I assume it should be `$results[] = ...`?

Reply:
MatchUrl changes $results. Changed this into a changeable variable now

Reply:
Don't like it very much, as the change of a variable is hidden a little bit, would prefer a clear return value. But for now it is okay, let's come to an end.